packages/timeline-state-resolver/src/integrations/websocketTcpClient/index.ts

```diff
@@ -0,0 +1,154 @@
+		this.connection = new WebSocketTcpConnection(options)
+        await this.connection.connect()
+        return true
+    }
+
+	readonly actions = {
+		[WebsocketTcpClientActions.Reconnect]: async (_id: string) => {
+			await this.connection.connect()
+			return { result: ActionExecutionResultCode.Ok }
+		},
+		[WebsocketTcpClientActions.ResetState]: async (_id: string) => {
+			return { result: ActionExecutionResultCode.Ok }
+		},
+		[WebsocketTcpClientActions.SendWebSocketMessage]: async (_id: string, payload?: Record<string, any>) => {
+			if (!payload?.message) {
+				return { result: ActionExecutionResultCode.Error, response: { key: 'Missing message in payload' } }
+			}
+			await this.sendCommand(payload.message)
+			return { result: ActionExecutionResultCode.Ok }
+		},
+		[WebsocketTcpClientActions.SendTcpMessage]: async (_id: string, payload?: Record<string, any>) => {
+			if (!payload?.command) {
+				return { result: ActionExecutionResultCode.Error, response: { key: 'Missing command in payload' } }
+			}
+			await this.sendCommand(payload.command)
+			return { result: ActionExecutionResultCode.Ok }
+		},
+	}
+
+	public get connected(): boolean {
+		return this.connection?.connected() ?? false
+	}
+
+	public getStatus(): Omit<DeviceStatus, "active"> {
+		return this.connection?.connectionStatus() ?? { statusCode: StatusCode.BAD, messages: ['No Connection'] }
+	}
+
+	public convertTimelineStateToDeviceState(
+		state: WebSocketTcpClientDeviceState 
+	): WebSocketTcpClientDeviceState {
+		// When a new Timeline State is about to be executed
+		// This is called to convert the generic Timeline State into a custom "device state".
+		// For example:
+		// tl obj: { layer: 'abc', content: { type: TimelineContentTypeWebSocketTcpClient::TCP_MESSAGE, message: 'hello'} }
+		// can be converted into: { tcpMessages: { abc: 'hello } }
+		//
+		// This is optional and for convenience only (like to simplify the diffing logic in diffStates())
+
+		return state
+	}
+
+	// ** Calculate Diffs of state and create the commands
+	public diffStates(oldState: WebSocketTcpClientDeviceState, newState: WebSocketTcpClientDeviceState): WebSocketTcpCommand[] {
+		// This is called to calculate and creates the commands needed to make olState reflect newState.
+		// Note: We DON'T send the commands NOW, but rather we return a list of the commands, to be executed
+		// later (send to sendCommand() ).
+
+
+		const commands: WebSocketTcpCommand[] = []
+		for (const [layerName, timelineObject] of Object.entries(newState.layers)) {
+			if (timelineObject.content.deviceType !== DeviceType.WEBSOCKET_TCP_CLIENT) continue
+
+
+			// We should send the command whenever the timeline object content has been ADDED or CHANGED
+			let changeType = 'N/A'
+			if (!oldState.layers[layerName]) {
+				changeType = 'added'
+			} else if (JSON.stringify(oldState.layers[layerName].content) !== JSON.stringify(timelineObject.content)) {
```

Comment:
`oldState` and `oldState.layers[layerName]` could be undefined
```suggestion
			} else if (JSON.stringify(oldState?.layers[layerName]?.content) !== JSON.stringify(timelineObject.content)) {
```